bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/ColumnViewer.java

```diff
@@ -817,4 +840,208 @@ protected void setBusy(boolean busy) {
+	 *                        already expanded by user.
+	 * @return list of children to be displayed/refreshed or {@code null} if given
+	 *         items length does not exceed items limit
+	 */
+	Object[] getChildrenWithLimitApplied(final Object parent, Item[] visibleChildren) {
+		final int limit = getItemsLimit();
+		final int visibleItemsLength = visibleChildren.length;
+		if (visibleItemsLength <= limit || limit <= 0) {
+			return null;
+		}
+		// if we are here, at least one iteration of expandable items is shown
+
+		Object[] currentSorted = getSortedChildren(parent);
+		// if current children count is below limit just return them
+		if (currentSorted.length < limit) {
+			return currentSorted;
+		}
+
+		// numberOfTimesRead is equal to number of times it was expanded by user.
+		int numberOfTimesRead = visibleItemsLength / limit;
+		if ((visibleItemsLength % limit) > 0
+				&& !(visibleChildren[visibleItemsLength - 1].getData() instanceof IExpandableNode)) {
+			numberOfTimesRead++;
+		}
+		List<Object> newChildren = new ArrayList<>();
+		Object lastElement = parent;
+		while (numberOfTimesRead > 0) {
+			Object[] sortedChildren;
+			if (lastElement == parent) {
+				sortedChildren = currentSorted;
+			} else {
+				sortedChildren = getSortedChildren(lastElement);
+			}
+			lastElement = sortedChildren[sortedChildren.length - 1];
+			Collections.addAll(newChildren, sortedChildren);
+
+			if (!(lastElement instanceof IExpandableNode)) {
+				// no more to read: less items than visibleChildren but more than limit.
+				break;
+			}
+			numberOfTimesRead--;
+			if (numberOfTimesRead > 0) {
+				// remove expandable node from the end of the list
+				// we will expand it at least once more on next loop round
+				newChildren.remove(newChildren.size() - 1);
+			}
+		}
+		return newChildren.toArray(new Object[newChildren.size()]);
+	}
+
+	/**
+	 * Returns the current viewer limit on direct children at one level. Limit that
+	 * is less than or equal to zero has no effect on the viewer.
+	 *
+	 * @return current limit
+	 * @since 3.31
+	 */
+	public int getItemsLimit() {
+		return itemsLimit;
+	}
+
+	/**
+	 * Sets the viewers items limit on direct children at one level.
+	 * <p>
+	 * If the number of direct children will exceed this limit, the viewer will only
+	 * show a subset of children up to the limit and add an {@link IExpandableNode}
+	 * element after last shown item.
+	 * </p>
+	 * <p>
+	 * This method must be called before {@link #setInput(Object)}. A parameter less
+	 * than or equal to zero has no effect on the viewer.
+	 * </p>
+	 * <p>
+	 * Client can adapt {@link ILabelProvider} to provide custom label and icon for
+	 * {@link IExpandableNode} element, otherwise default label and no icon will be
+	 * used.
+	 * </p>
+	 *
+	 * @param newLimit A non-negative integer greater than 0 to enable items limit
+	 * @since 3.31
+	 */
+	public void setItemsLimit(int newLimit) {
```

Comment:
I don't think the name is best here. A "limit" gives the impression we cannot go beyond. Here what we want to control is more the "size" of an "increment" in displaying the output. I would suggest a name such as `setDisplayIncrementally(int incrementSize)`

Reply:
fixed.